### PR TITLE
Upgrades to pycbc_grb_inj_finder

### PR DESCRIPTION
**Pull request by ETVincent**
retain single ifo injection information and store in hdf output file

### CODE REVIEW

File: bin/pygrb/pycbc_grb_inj_finder [hunk 1]
```diff
@@ -83,8 +82,7 @@ def read_segment_files(segfiles):
         map(_read, segfiles),
         segmentlist()))
 
-
-def read_hdf5_triggers(inputfiles, verbose=False):
+def read_hdf5_triggers(inputfiles, trig_dict, verbose=False):
     """Merge several HDF5 files into a single file
 
     Parameters
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:88]: Please state what `trig_dict` is

File: bin/pygrb/pycbc_grb_inj_finder [hunk 2]
```diff
@@ -95,56 +93,101 @@ def read_hdf5_triggers(inputfiles, verbose=False):
...
+    
+    out_dict = {}
+    for x in trig_dict:
+        out_dict[x] = {}
+        
+    placeholder = 0
+
+    # create datasets    
+    for dset, (shape, dtype) in datasets_ifo.items():
+        # 3: chops off H1/ or L1/ or V1/ from the dset string
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:154]: Please add `K1/` to the list

File: bin/pygrb/pycbc_grb_inj_finder [hunk 3]
```diff
@@ -95,56 +93,101 @@ def read_hdf5_triggers(inputfiles, verbose=False):
...
-
-    out = {}
-
-    # create datasets
-    for dset, (shape, dtype) in datasets.items():
-        # 8: chops off network/ from the dset string
-        out[dset[8:]] = numpy.empty(shape, dtype=dtype)
+    
+    out_dict = {}
+    for x in trig_dict:
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:147]: It should be possible to copy the `keys()` in a more pythonic way.

File: bin/pygrb/pycbc_grb_inj_finder [hunk 4]
```diff
@@ -95,56 +93,101 @@ def read_hdf5_triggers(inputfiles, verbose=False):
...
+    for dset, (shape, dtype) in datasets_ifo.items():
+        # 3: chops off H1/ or L1/ or V1/ from the dset string
+        for x in trig_dict:
+            if placeholder == 0 and x!= 'network':
+                placeholder = x
+            if dset.startswith(x):
+                out_dict[x][dset[3:]] = numpy.empty(shape, dtype=dtype)
+            
+    for dset, (shape, dtype) in datasets_n.items():
+        out_dict['network'][dset[8:]] = numpy.empty(shape, dtype=dtype)
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:162]: As you did for `H1/`, etc. please add a comment to remind us that the 8 skips `network/`

File: bin/pygrb/pycbc_grb_inj_finder [hunk 5]
```diff
@@ -95,56 +93,101 @@ def read_hdf5_triggers(inputfiles, verbose=False):
...
+                        
+                for dset in datasets_n: 
+                    data = h5in[dset][:]
+                    size = data.shape[0]
+                    pos = position[dset]
+                    if EVENT_ID_REGEX.search(dset):
+                        out_dict['network'][dset[8:]][pos:pos+size] = data + pos
+                    else:
+                        out_dict['network'][dset[8:]][pos:pos+size] = data
+                    position[dset] += size
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:188]: It should be possible to turn these 2 for loops into 2 calls to a single function (with appropriate arguments)

File: bin/pygrb/pycbc_grb_inj_finder [hunk 6]
```diff
@@ -228,7 +271,16 @@ exclude = read_segment_files(args.exclude_segments)
...
+
+# dictionary that will contain ifo names and 'network' 
+ifo_trigs_dict = {}
+# open first injection file
+firstfile = h5py.File(args.input_files[0], "r")
+for key in firstfile.keys():
+    # populate ifo_trigs_dict with ifo names and 'network' 
+    ifo_trigs_dict[key] = defaultdict(list)
+    
+print(ifo_trigs_dict)
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:283]: This is for debugging purposes and can be removed.

File: bin/pygrb/pycbc_grb_inj_finder [hunk 7]
```diff
@@ -251,14 +303,18 @@ with tqdm.tqdm(args.inj_files, desc="Finding injections",
         nexcluded += (~keep).sum()
         injections = tmpinj[keep]
         injtime = injtime[keep]
-
+
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:306]: Remove whitespaces

File: bin/pygrb/pycbc_grb_inj_finder [hunk 8]
```diff
@@ -251,14 +303,18 @@ with tqdm.tqdm(args.inj_files, desc="Finding injections",
...
+        
         # read triggers
-        triggers = read_hdf5_triggers(trigfiles)
-        time = triggers["end_time_gc"]
-        snr = triggers[args.rank_column]
-        event_id = triggers["event_id"]
+        triggers_dict = read_hdf5_triggers(trigfiles, ifo_trigs_dict)
+        
+        #print(triggers_dict['network'].keys())
+        #print(ifotriggers.keys())
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:311]: Remove these lines

File: bin/pygrb/pycbc_grb_inj_finder [hunk 9]
```diff
@@ -251,14 +303,18 @@ with tqdm.tqdm(args.inj_files, desc="Finding injections",
...
+        
+        time = triggers_dict['network']['end_time_gc']
+        snr = triggers_dict['network'][args.rank_column]
+        event_id = triggers_dict['network']["event_id"]
         time_sorting = time.argsort()
-
+
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:317]: Remove whitespaces

File: bin/pygrb/pycbc_grb_inj_finder [hunk 10]
```diff
@@ -321,15 +378,18 @@ with h5py.File(outfilename, "w") as h5out:
...
-        netg.create_dataset(
-            col,
-            data=trigs[col],
-            compression="gzip",
-            compression_opts=9,
-        )
-
+    
+    for x in ifo_trigs_dict:
+        next = h5out.create_group(x)
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:385]: Maybe use `x_group` rather than `next` in these lines

**Review by pannarale — CHANGES_REQUESTED** [review]

File: bin/pygrb/pycbc_grb_inj_finder [hunk 11]
```diff
@@ -133,6 +136,32 @@ def read_hdf5_triggers(inputfiles, trig_dict, verbose=False):
                 ))
         datasets_n[name] = (shape, dtype)
 
+    def cp_dataset_cont(dataset, xlen, xidentifier, out_dct, pos_full):
+        """Copy dataset contnets into out dictionary.
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:140]:
```suggestion
        """Copy dataset contents into output dictionary (out_dct).
```

File: bin/pygrb/pycbc_grb_inj_finder [hunk 12]
```diff
@@ -133,6 +136,32 @@ def read_hdf5_triggers(inputfiles, trig_dict, verbose=False):
...
+
+        Parameters
+        ----------
+        dataset : iteration of datasets_ifo or datasets_n
+
+        xlen : int. The length of IFO identifier (3) or network (8).
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:146]: Given `xidentifier` let the function determine `xlen`, so there is no need to pass `xlen`

File: bin/pygrb/pycbc_grb_inj_finder [hunk 13]
```diff
@@ -133,6 +136,32 @@ def read_hdf5_triggers(inputfiles, trig_dict, verbose=False):
...
+
+        Parameters
+        ----------
+        dataset : iteration of datasets_ifo or datasets_n
+
+        xlen : int. The length of IFO identifier (3) or network (8). 
+
+        xidentifier : IFO identifier (ex. 'H1') or 'network'.
+
+        out_dct: out_dict to be modified
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:150]:
```suggestion
        out_dct: output dictionary to be modified
```

**Review by pannarale — CHANGES_REQUESTED** [review]

File: bin/pygrb/pycbc_grb_inj_finder (hunk 12 above)
Comment (ETVincent) [reply bin/pygrb/pycbc_grb_inj_finder]: Addressed in recent file update.

**Review by pannarale — APPROVED** [review]

**Comment by ETVincent** [comment]
Bug found. Do not pull yet.

**Comment by ETVincent** [comment]
> Bug found. Do not pull yet.

Corrected in new commit.

File: bin/pygrb/pycbc_grb_inj_finder [hunk 14]
```diff
@@ -84,67 +83,69 @@ def read_segment_files(segfiles):
...
 
     Parameters
     ----------
     inputfiles : `list` of `str`
         the paths of the input HDF5 files to merge
 
-    outputfile : `str`
-        the path of the output HDF5 file to write
+    trig_dict: dictionary 
+        dictionary with IFO or network identifier as keys
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:95]: It looks like `trig_dict` is no longer used: please remove here and from the function's arguments.

File: bin/pygrb/pycbc_grb_inj_finder [hunk 15]
```diff
@@ -84,67 +83,69 @@ def read_segment_files(segfiles):
...
+            else:
+                assert dtype == datasets[name][1], (
+                    "Cannot merge {0}/{1}, does not match dtype".format(
+                        obj.file.filename, name,
+                    ))
+            datasets[name] = (shape, dtype)
 
     # get list of datasets
-    datasets = {}
+    #datasets = {}
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:120]: Remove.

File: bin/pygrb/pycbc_grb_inj_finder [hunk 16]
```diff
@@ -253,10 +262,10 @@ with tqdm.tqdm(args.inj_files, desc="Finding injections",
         injtime = injtime[keep]
 
         # read triggers
-        triggers = read_hdf5_triggers(trigfiles)
-        time = triggers["end_time_gc"]
-        snr = triggers[args.rank_column]
-        event_id = triggers["event_id"]
+        triggers = read_hdf5_triggers(trigfiles, out_trigs)
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:265]: And then here you won't need `out_trigs`.

**Review by pannarale — CHANGES_REQUESTED** [review]

File: bin/pygrb/pycbc_grb_inj_finder [hunk 17]
```diff
@@ -117,7 +115,7 @@ def read_hdf5_triggers(inputfiles, trig_dict, verbose=False):
             datasets[name] = (shape, dtype)
 
     # get list of datasets
-    #datasets = {}
+    datasets = {}
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:118]: The line needs to be removed.

**Review by pannarale — CHANGES_REQUESTED** [review]

File: bin/pygrb/pycbc_grb_inj_finder [hunk 18]
```diff
@@ -92,28 +91,28 @@ def read_hdf5_triggers(inputfiles, verbose=False):
     inputfiles : `list` of `str`
         the paths of the input HDF5 files to merge
```
Comment (pannarale) [inline bin/pygrb/pycbc_grb_inj_finder:93]: This empty line should be removed

**Review by pannarale — CHANGES_REQUESTED** [review]

**Review by pannarale — APPROVED** [review]

**Pull request merged**